22.md

```diff
@@ -0,0 +1,179 @@
+NIP-22
+======
+
+Comment
+-------
+
+`draft` `optional`
+
+A comment is a threading note always scoped to a root event or an `S`-tag.
+
+It uses `kind:1111` with plaintext `.content` (no HTML, Markdown, or other formatting).
+
+Comments MUST point to the root scope using uppercase tag names (e.g. `K`, `E`, `A` or `S`)
+and MUST point to the parent item with lowercase ones (e.g. `k`, `e`, `a` or `s`).
+
+```js
+{
+  kind: 1111,
+  content: '<comment>',
+  tags: [
+    // root scope: event addresses, event ids, or S-tags.
+    ["<A, E, S>", "<address, id or S-value>", "<relay or web page hint>", "<root event's pubkey, if an E tag>"],
+    // the root kind or type if r.
+    ["K", "<root kind>"],
+
+    // parent item: event addresses, event ids, or s-tags.
+    ["<a, e, s>", "<address, id or s-value>", "<relay or web page hint>", "<parent event's pubkey, if an e tag>"],
+    // parent item kind
+    ["k", "<parent comment kind>"]
+  ]
+  // other fields
+}
+```
+
+Tags `K` and `k` MUST be present to define the event kind of the root and the parent items.
+
+`S` and `s` tags create scopes for hashtags, geohashes, URLs, and other external identifiers. They accept the following values:
+
+| Type | `S`/`s` tag | `K`/`k` tag |
+|- | - | - |
+| URLs | "r:`<URL, normalized, no fragment>`" | "r:`<scheme-host, normalized>`" |
+| Hashtags | "t:`<topic, lowercase>`" | "#:t" |
+| Geohashes| "g:`<geohash, lowercase>`" | "#:g" |
+| Podcast Feeds | "i:podcast:guid:`<id, lowercase>`" | "i:podcast:guid" |
+| Podcast Episodes | "i:podcast:item:guid:`<id, lowercase>`" | "i:podcast:item:guid" |
+| Books | "i:isbn:`<id, hyphenated>`" | "i:isbn" |
+| Movies | "i:isan:`<id, lowercase>`" | "i:isan" |
+| Papers | "i:doi:`<id, lowercase>`" | "i:doi" |
+
+## Examples
+
+A comment on a blog post looks like this:
+
+```js
+{
+  kind: 1111,
+  content: 'Great blog post!',
+  tags: [
+    // top-level comments scope to event addresses or ids
+    ["A", "30023:3c9849383bdea883b0bd16fece1ed36d37e37cdde3ce43b17ea4e9192ec11289:f9347ca7", "wss://example.relay"],
+    // the root kind
+    ["K", "30023"],
+
+    // the parent event address (same as root for top-level comments)
+    ["a", "30023:3c9849383bdea883b0bd16fece1ed36d37e37cdde3ce43b17ea4e9192ec11289:f9347ca7", "wss://example.relay"],
+    // when the parent event is replaceable or addressable, also include an `e` tag referencing its id
+    ["e", "5b4fc7fed15672fefe65d2426f67197b71ccc82aa0cc8a9e94f683eb78e07651", "wss://example.relay"],
```

Comment:
Don't references inside the content use the tag `q`?

Reply:
Can we use `q` with addresses and not only event ids? @jb55?

Reply:
Maybe do id,relay,empty,address?

`["q", "231e65b29f063ea81eb7435e61f93b4a4429a6a159e432bcdcbad55d8896bcec", "wss://relay.whatever", "", "30023:79be667ef9dcbbac55a06295ce870b07029bfcdb2dce28d959f2815b16f81798:banana"]`

Reply:
You won't be able to query by address, then.

Reply:
I think you missed what they @dluvian was saying, @arthurfranca.

On a kind 1111, users should be able to mention any other event including the already tagged A/E/a/e. If we add another `e` it will look like `1111` is replying to both events. So I don't think that is what we want. Only one `e` MUST be present. 

For all mentions, of all kinds, inside the content, including mentioning the root and parent directly, a `q` tag SHOULD be added together with the `nevent` and `naddr` links in the text itself. 

The format
```
["q", <event-id> or <event-address>, <relay-url>, <pubkey-if-a-regular-event>]
```
makes sense to me. But it has never been merged into this repo.

Reply:
I like the `q` tag. I forgot that it exists. Going over the main README, I find nothing that describes quotes/references. It's kind of hidden under the `Reposts` nip.

Reply:
Maybe we should use this event to start the magic `q` tag with both formats.

Reply:
> I like the `q` tag. I forgot that it exists. Going over the main README, I find nothing that describes quotes/references. It's kind of hidden under the `Reposts` nip.

Yeah, the `q` tag was forced upon us and prematurely merged without addressing neither its issues nor adapting the other NIPs to use it instead of the `mention` marker. Currently we have a big mess of the two formats.

Reply:
> But wouldn't nostr:nevent1 without extra tags be enough? @dluvian


If I comment on anything and mention git issue B, I would like to show the referencing comment in issue B. Just like in GitHub. So a tag is needed and I don't want to overload the `e` tag again.